ext/strings.go

```diff
@@ -485,6 +549,490 @@ func join(strs []string) (string, error) {
+			}
+		}
+	}
+	_, err = listStrBuilder.WriteRune(']')
+	if err != nil {
+		return "", fmt.Errorf("error writing to list string: %w", err)
+	}
+	return listStrBuilder.String(), nil
+}
+
+func formatMap(arg ref.Val, locale string) (string, error) {
+	argMap := arg.(traits.Mapper)
+	argIterator := argMap.Iterator()
+	type mapPair struct {
+		key   string
+		value string
+	}
+	argPairs := make([]mapPair, argMap.Size().Value().(int64))
+	i := 0
+	for argIterator.HasNext() == types.True {
+		key := argIterator.Next()
+		var keyFormat clauseImpl
+		switch key.Type() {
+		case types.StringType:
+			keyFormat = formatString
+		case types.IntType, types.UintType:
+			keyFormat = formatDecimal
+		case types.BoolType:
+			keyFormat = formatBool
+		default:
+			return "", fmt.Errorf("no formatting function for map key of type %s", key.Type().TypeName())
+		}
+		unquotedKeyStr, err := keyFormat(key, locale)
+		if err != nil {
+			return "", err
+		}
+		keyStr := quoteForCEL(key, unquotedKeyStr)
+		value, found := argMap.Find(key)
+		if !found {
+			return "", fmt.Errorf("could not find key: %q", key)
+		}
+		valueFormat, err := clauseForType(value.Type())
+		if err != nil {
+			return "", err
+		}
+		unquotedValueStr, err := valueFormat(value, locale)
+		if err != nil {
+			return "", err
+		}
+		valueStr := quoteForCEL(value, unquotedValueStr)
+		argPairs[i] = mapPair{keyStr, valueStr}
+		i++
+	}
+	sort.SliceStable(argPairs, func(x, y int) bool {
+		return argPairs[x].key < argPairs[y].key
+	})
+	var mapStrBuilder strings.Builder
+	_, err := mapStrBuilder.WriteRune('{')
+	if err != nil {
+		return "", fmt.Errorf("error writing to map string: %w", err)
+	}
+	for i, entry := range argPairs {
+		_, err = mapStrBuilder.WriteString(fmt.Sprintf("%s:%s", entry.key, entry.value))
+		if err != nil {
+			return "", fmt.Errorf("error writing to map string: %w", err)
+		}
+		if i < len(argPairs)-1 {
+			_, err = mapStrBuilder.WriteString(", ")
+			if err != nil {
+				return "", fmt.Errorf("error writing to map string: %w", err)
+			}
+		}
+	}
+	_, err = mapStrBuilder.WriteRune('}')
+	if err != nil {
+		return "", fmt.Errorf("error writing to map string: %w", err)
+	}
+	return mapStrBuilder.String(), nil
+}
+
+// quoteForCEL takes a formatted, unquoted value and quotes it in a manner
+// suitable for embedding directly in CEL.
+func quoteForCEL(refVal ref.Val, unquotedValue string) string {
+	switch refVal.Type() {
+	case types.StringType:
+		return fmt.Sprintf("%q", unquotedValue)
+	case types.BytesType:
+		return fmt.Sprintf("b%q", unquotedValue)
+	case types.DoubleType:
+		// special case to handle infinity/NaN
+		num := refVal.Value().(float64)
+		if math.IsInf(num, 1) || math.IsInf(num, -1) || math.IsNaN(num) {
+			return fmt.Sprintf("%q", unquotedValue)
+		}
+		return unquotedValue
+	default:
+		return unquotedValue
+	}
+}
+
+func formatString(arg ref.Val, locale string) (string, error) {
+	switch arg.Type() {
```

Comment:
Note: all of the simple types in CEL have conversions to string, so it's possible you could attempt conversion to string for any simple `ref.Val` value.

Reply:
That's definitely true, though wouldn't we lose out on potential type-safety (i.e, passing a number to `%d` or `%x` instead)? @jpbetz had wanted (and I was going to implement as a follow-up PR) compile-time checks to make sure the types passed to `string.format` were correct when the expression allows for doing so - I think we talked about that here somewhere.

Reply:
That's a fair point, but if you provide a `list` to `%s` and it contains all numeric values, what's the expected behavior?

You can either go toward restricting the types to make sure they're used appropriately (true for all numeric formatting clauses) or more permissive and allowing as much as possible which can be converted to that formatting clause into a human-readable string. For `%s` I'm inclined to support everything, but can see preferring to direct users toward the numeric formatting controls in those cases since they offer better fidelity / clarity to the person reviewing the code.

Reply:
When passing a list to `%s` and there's only numeric values, `%d` is used to treat the integers and `%f` is used to treat doubles; I think that's a decent middle ground - for type-checking it's tricky to be as thorough as with scalar types since they can contain more than one type as an element. We could have a dedicated list/map clause, but I think that might be surprising coming from other languages. 

It might follow that adding a dedicated bool clause/adding `%t` back would also make sense, though I think the behavior with `%s` and `%t` right now is fine (though I wouldn't push back if others wanted to go back the `%t` route), and the other types that `%s` supports (like duration/timestamp) are probably too specific to warrant their own formatting clause.

Reply:
>  For %s I'm inclined to support everything.

The current state of `%s` has been bothering me for a while and I should have flagged it earlier too. My complaint is that it supports selected types, but if you asked a casual developer which types it supports and why, I don't see how they could figure out the rhyme or reason of the types we've shoehorned into to `%s`.

I think we should either restrict `%s` to only string, or allow everything. If we do restrict `%s` to only string, then at the very least need to provide a way to "data dump" lists and maps. This could be provided either with a `%v` type directive or CEL utility functions. If we allow everything, we need to document the behavior (for most types, I would expect that the behavior matches a direct `string(whatever)` conversion.

Reply:
Right - I've added numerical support to `%s`, which is equivalent to everything (it was already used for `Timestamp`/`Duration` before, etc.).